python/label_pr.py

```diff
@@ -0,0 +1,153 @@
+    """
+
+    org, repo, issue_num = get_owner_repo_num(html_url)
+    return gh.get_organization(org).get_repo(repo).get_issue(issue_num)
+
+
+def get_pull_request(gh: Github, html_url: str) -> PullRequest:
+    """
+    Get the ``PullRequest`` instance from a GitHub PR's HTML URL.
+
+    :param gh: the GitHub client
+    :param html_url: the HTML URL of the PR
+    :return: the ``PullRequest`` instance corresponding to the given URL
+    """
+
+    org, repo, pr_num = get_owner_repo_num(html_url)
+    return gh.get_organization(org).get_repo(repo).get_pull(pr_num)
+
+
+def get_linked_issues(url: str) -> list[str]:
+    """
+    Get the list of linked issues from the GitHub UI by parsing the HTML. This is a
+    workaround because GitHub API does not provide a reliable way to find the linked
+    issues for a PR.
+
+    :param url: the URL to the GitHub UI for a PR
+    :return: the list of HTML URLs for linked issues
+    """
+
+    res = requests.get(url)
+    if res.status_code != 200:
+        return []
+
+    soup = BeautifulSoup(res.text, "html.parser")
+    divs = soup.find_all("div", **{"class": re.compile("css-truncate my-1")})
+    anchors = [div.a["href"] for div in divs]
+    return anchors
+
+
+def get_label_of_cat(cat: str, labels: list[Label]) -> Optional[Label]:
+    """
+    Get the label of a particular category from the given list of labels.
+
+    :param cat: the category to which the label should belong
+    :param labels: the list of labels to choose from
+    :return: the label matching the given category
+    """
+
+    for label in labels:
+        if cat in label.name:
+            return label
+    return None
+
+
+def main():
+    configure_logger()
+
+    args = parser.parse_args()
+
+    log.debug(f"PR URL: {args.pr_url}")
+
+    github_info = get_data("github.yml")
+    org_handle = github_info["org"]
+    log.info(f"Organization handle: {org_handle}")
+
+    gh = get_client()
+
+    pr = get_pull_request(gh, args.pr_url)
+    log.info(f"Found PR: {pr.title}")
+
+    linked_issues = get_linked_issues(pr.html_url)
+    log.info(f"Found {len(linked_issues)} linked issues")
+
+    is_labelled = False
+    if linked_issues:
+        for issue in linked_issues:
+            issue = get_issue(gh, issue)
+            labels = issue.labels
+            if all(
+                [
+                    (aspect := get_label_of_cat("aspect", labels)),
+                    (priority := get_label_of_cat("priority", labels)),
+                    (goal := get_label_of_cat("goal", labels)),
+                ]
+            ):
+                log.info(f"Aspect label: {aspect}")
+                log.info(f"Priority label: {priority}")
+                log.info(f"Goal label: {goal}")
+                pr.set_labels(aspect, priority, goal)
+                is_labelled = True
+                break
+
+    if not is_labelled:
+        log.info("Could not find properly labelled issue")
+        pr.set_labels("🚦 status: awaiting triage")
```

Comment:
I _think_ this could be replaced with a [`for...else` loop](https://book.pythontips.com/en/latest/for_-_else.html) (we may need to default `linked_issues` to an empty list)

```suggestion
    for issue in linked_issues:
        issue = get_issue(gh, issue)
        labels = issue.labels
        if all(
            [
                (aspect := get_label_of_cat("aspect", labels)),
                (priority := get_label_of_cat("priority", labels)),
                (goal := get_label_of_cat("goal", labels)),
            ]
        ):
            log.info(f"Aspect label: {aspect}")
            log.info(f"Priority label: {priority}")
            log.info(f"Goal label: {goal}")
            pr.set_labels(aspect, priority, goal)
            break
    else:
        log.info("Could not find properly labelled issue")
        pr.set_labels("🚦 status: awaiting triage")
```

Reply:
Great use of `for`-`else`. Thanks!